plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/conversions/RemoveUnnecessaryParenthesesConversion.kt

```diff
@@ -0,0 +1,213 @@
+        if ((parent is JKLocalVariable && parent.initializer == expression) ||
+            (parent is JKField && parent.initializer == expression) ||
+            (parent is JKKtAssignmentStatement && parent.expression == expression)
+        ) {
+            return false
+        }
+
+
+        if (parent is JKPrefixExpression || parent is JKPostfixExpression) {
+            return !isAtomic(innerExpression)
+        }
+
+        // remove parentheses if doing so doesn't affect the order of operations
+        val innerPriority = getPriority(innerExpression)
+        val parentPriority = getPriority(parent)
+        if (innerPriority == parentPriority) {
+            if (parent !is JKBinaryExpression) return false
+            if (innerExpression is JKBinaryExpression &&
+                (innerExpression.operator.token == JKOperatorToken.ANDAND ||
+                        innerExpression.operator.token == JKOperatorToken.OROR)
+            ) {
+                return false
+            }
+            return parent.right == expression
+        }
+
+        return innerPriority < parentPriority
+    }
+
+    private fun isAtomic(expression: JKExpression): Boolean {
+        return expression is JKQualifiedExpression ||
+                expression is JKKtWhenExpression ||
+                expression is JKCallExpression ||
+                expression is JKFieldAccessExpression ||
+                expression is JKLiteralExpression
+    }
+
+    private fun recursivelyContainsNewlines(expression: JKExpression): Boolean {
+        if (expression.children.isEmpty()) return false
+        if (expression.children.filterIsInstanceAnd<JKExpression> { it.containsNewLine() }.isNotEmpty()) return true
+        return expression.children.filterIsInstance<JKExpression>().any { recursivelyContainsNewlines(it) }
+    }
+
+    private fun recursivelyContainsNewlineBeforeOperator(expression: JKBinaryExpression): Boolean {
+        val left = expression.left
+        val right = expression.right
+        if (recursivelyEndsWithNewline(left)) {
+            return true
+        }
+        if (left is JKBinaryExpression && recursivelyContainsNewlineBeforeOperator(left)) return true
+        if (right is JKBinaryExpression && recursivelyContainsNewlineBeforeOperator(right)) return true
+        return false
+    }
+
+    private fun recursivelyEndsWithNewline(expression: JKExpression): Boolean {
+        if (expression.hasLineBreakAfter) return true
+        val lastChild = expression.children.lastOrNull()?.safeAs<JKExpression>() ?: return false
+        return recursivelyEndsWithNewline(lastChild)
+    }
+
+    private fun recursivelyStartsWithNewline(expression: JKExpression): Boolean {
+        if (expression.hasLineBreakBefore) return true
+        val lastChild = expression.children.firstOrNull()?.safeAs<JKExpression>() ?: return false
+        return recursivelyStartsWithNewline(lastChild)
+    }
+
+    /**
+     * Based on `getPriority` in KtPsiUtil.java
```

Comment:
Adding this reference is great, but using the FQN is even better: `org.jetbrains.kotlin.psi.KtPsiUtil#getPriority`.